File: app/src/main/java/mega/privacy/android/app/listeners/QueryRecoveryLinkListener.java
package mega.privacy.android.app.listeners;

import static mega.privacy.android.app.utils.Constants.ACTION_PARK_ACCOUNT;
import static mega.privacy.android.app.utils.Constants.ACTION_RESET_PASS;
import static mega.privacy.android.app.utils.Constants.RESET_PASSWORD_LINK_REGEXS;
import static mega.privacy.android.app.utils.Constants.TOUR_FRAGMENT;
import static mega.privacy.android.app.utils.Constants.VISIBLE_FRAGMENT;
import static mega.privacy.android.app.utils.Util.matchRegexs;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import mega.privacy.android.app.OpenLinkActivity;
import mega.privacy.android.app.R;
import mega.privacy.android.app.main.LoginActivity;
import nz.mega.sdk.MegaApiJava;
import nz.mega.sdk.MegaError;
import nz.mega.sdk.MegaRequest;
import timber.log.Timber;

public class QueryRecoveryLinkListener extends BaseListener {

    OpenLinkActivity openLinkActivity;

    public QueryRecoveryLinkListener(Context context) {
        super(context);
    }

    @Override
    public void onRequestFinish(MegaApiJava api, MegaRequest request, MegaError e) {
        if (request.getType() != MegaRequest.TYPE_QUERY_RECOVERY_LINK) return;
        Timber.d("TYPE_GET_RECOVERY_LINK");

        if (context instanceof OpenLinkActivity) {
            openLinkActivity = (OpenLinkActivity) context;
        }

        String url = request.getLink();

        if (openLinkActivity != null) {
            if (url == null) {
                Timber.w("Error opening link URL null: %s___%d", e.getErrorString(), e.getErrorCode());
                openLinkActivity.setError(openLinkActivity.getString(R.string.general_text_error));
            }

            switch (e.getErrorCode()) {
                case MegaError.API_OK:
                    Intent intent = null;

                    if (matchRegexs(url, RESET_PASSWORD_LINK_REGEXS)) {
                        intent = new Intent(openLinkActivity, LoginActivity.class);
                        intent.putExtra(VISIBLE_FRAGMENT, TOUR_FRAGMENT);
                        if (request.getFlag()) {
                            intent.setAction(ACTION_RESET_PASS);
                        } else {
                            intent.setAction(ACTION_PARK_ACCOUNT);
                        }
                    }

                    if (intent != null) {
                        intent.setData(Uri.parse(url));
                        openLinkActivity.startActivity(intent);
                        openLinkActivity.finish();
                    }
                    break;

                case MegaError.API_EEXPIRED:
                    if (matchRegexs(url, RESET_PASSWORD_LINK_REGEXS)) {
                        openLinkActivity.setError(openLinkActivity.getString(R.string.recovery_link_expired));
                    }
                    break;

                case MegaError.API_EACCESS:
                    if (matchRegexs(url, RESET_PASSWORD_LINK_REGEXS)) {
                        Timber.w("Error opening link not related to this account: %s___%d", e.getErrorString(), e.getErrorCode());
                        openLinkActivity.setError(openLinkActivity.getString(R.string.error_not_logged_with_correct_account));
                    }
                    break;

                default:
                    openLinkActivity.setError(openLinkActivity.getString(R.string.invalid_link));

            }
        }
    }
}

File: app/src/main/java/mega/privacy/android/app/listeners/QueryRecoveryLinkListener.kt
package mega.privacy.android.app.listeners

import android.content.Intent
import android.net.Uri
import mega.privacy.android.app.OpenLinkActivity
import mega.privacy.android.app.R
import mega.privacy.android.app.presentation.login.LoginActivity
import mega.privacy.android.app.utils.Constants
import mega.privacy.android.app.utils.Util
import nz.mega.sdk.MegaApiJava
import nz.mega.sdk.MegaError
import nz.mega.sdk.MegaRequest
import nz.mega.sdk.MegaRequestListenerInterface
import timber.log.Timber

/**
 * Listener to handle query recovery response
 *
 * @param activity : OpenLinkActivity
 */
class QueryRecoveryLinkListener(
    private val activity: OpenLinkActivity,
) : MegaRequestListenerInterface {

    /**
     * Callback function for onRequestStart
     *
     * @param api : MegaApiJava
     * @param request : MegaRequest
     */
    override fun onRequestStart(api: MegaApiJava, request: MegaRequest) {
        // Do nothing
    }

    /**
     * Callback function for onRequestUpdate
     *
     * @param api : MegaApiJava
     * @param request : MegaRequest
     */
    override fun onRequestUpdate(api: MegaApiJava, request: MegaRequest) {
        // Do nothing
    }

    /**
     * Callback function for onRequestFinish
     *
     * @param api : MegaApiJava
     * @param request : MegaRequest
     */
    override fun onRequestFinish(api: MegaApiJava, request: MegaRequest, e: MegaError) {
        if (request.type == MegaRequest.TYPE_QUERY_RECOVERY_LINK) {
            with(activity) {
                val url = request.link
                if (url.isNullOrEmpty()) {
                    Timber.w("Error opening link URL null: %s___%d", e.errorString, e.errorCode)
                    this.setError(this.getString(R.string.general_text_error))
                    return
                }

                when (e.errorCode) {
                    MegaError.API_OK -> {
                        if (Util.matchRegexs(url, Constants.RESET_PASSWORD_LINK_REGEXS)) {
                            Intent(this, LoginActivity::class.java).apply {
                                putExtra(Constants.VISIBLE_FRAGMENT, Constants.TOUR_FRAGMENT)
                                action = if (request.flag) {
                                    Constants.ACTION_RESET_PASS
                                } else {
                                    Constants.ACTION_PARK_ACCOUNT
                                }
                                data = Uri.parse(url)
                            }.run {
                                startActivity(this)
                                finish()
                            }
                        }
                    }

                    MegaError.API_EEXPIRED ->
                        if (Util.matchRegexs(url, Constants.RESET_PASSWORD_LINK_REGEXS)) {
                            this.setError(this.getString(R.string.recovery_link_expired))
                        }

                    MegaError.API_EACCESS ->
                        if (Util.matchRegexs(url, Constants.RESET_PASSWORD_LINK_REGEXS)) {
                            Timber.w(
                                "Error opening link not related to this account: %s___%d",
                                e.errorString,
                                e.errorCode)
                            this.setError(this.getString(R.string.error_not_logged_with_correct_account))
                        }

                    else -> this.setError(this.getString(R.string.invalid_link))
                }
            }
        }
    }

    /**
     * Callback function for onRequestTemporaryError
     *
     * @param api : MegaApiJava
     * @param request : MegaRequest
     */
    override fun onRequestTemporaryError(api: MegaApiJava, request: MegaRequest, e: MegaError) {
        // Do nothing
    }
}